wcfsetup/install/files/lib/data/user/avatar/UserAvatarEditor.class.php

```
@@ -66,11 +68,81 @@ public function deleteFiles()
+            @\unlink($this->getLocation(null, true));
+        }
+    }
+
+    /**
+     * Creates a WebP variant of the avatar, unless it is a GIF image. If the
+     * user uploads a WebP image, this method will create a JPEG variant as a
+     * fallback for ancient clients.
+     * 
+     * Will return `true` if a variant has been created.
+     *
+     * @since 5.4
+     */
+    public function createAvatarVariant(): bool
+    {
+        if ($this->hasWebP) {
+            return false;
+        }
+
+        if ($this->avatarExtension === "gif") {
+            // We do not touch GIFs at all.
+            return false;
+        }
+
+        $filename = $this->getLocation();
+        $filenameWebP = $this->getLocation(null, true);
+
+        $imageAdapter = ImageHandler::getInstance()->getAdapter();
+        $imageAdapter->loadFile($filename);
+        $image = $imageAdapter->getImage();
+
+        $data = ["hasWebP" => 1];
+
+        // If the uploaded avatar is already a WebP image, then create a JPEG
+        // as a fallback image and flip the image data to match the JPEG.
+        if ($this->avatarExtension === "webp") {
+            $filenameJpeg = preg_replace('~\.webp$~', '.jpeg', $filenameWebP);
+
+            $imageAdapter->saveImageAs($image, $filenameJpeg, "jpeg", 80);
+
+            // The new file has a different SHA1 hash, which means that apart from
+            // updating the `fileHash` we also need to move it to a different physical
+            // location.
+            $newFileHash = sha1_file($filenameJpeg);
+
+            $tmpAvatar = clone $this;
+            $tmpAvatar->data["avatarExtension"] = "jpeg";
+            $tmpAvatar->data["fileHash"] = $newFileHash;
+            $newLocation = $tmpAvatar->getLocation(null, false);
+
+            $dir = \dirname($newLocation);
+            if (!@\file_exists($dir)) {
+                FileUtil::makePath($dir);
+            }
+
+            @\rename($filenameJpeg, $newLocation);
```

And neither should it here. If this `rename` fails we must not silently fail.